compiler/rustc_lint/src/reachable_patterns.rs

```diff
@@ -0,0 +1,152 @@
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// When a struct or enum is marked as `non_exhaustive` it may still be useful to warn
+    /// when there are patterns that are not considered. This lint will catch these cases.
+    REACHABLE_PATTERNS,
+    Allow,
+    "detect when patterns of types marked `non_exhaustive` are missed",
+}
+
+declare_lint_pass!(ReachablePattern => [REACHABLE_PATTERNS]);
+
+impl<'tcx> LateLintPass<'tcx> for ReachablePattern {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) {
+        if let ExprKind::Match(
+            e @ Expr { kind: ExprKind::Path(..), .. },
+            arms,
+            MatchSource::Normal,
+        ) = &expr.kind
+        {
+            if !arms.is_empty() {
+                // TODO: can this ever not be there?
+                if let Some(ty::Adt(def, _)) =
+                    cx.maybe_typeck_results().map(|r| r.expr_ty(e).kind())
+                {
+                    if def.is_enum() && def.is_variant_list_non_exhaustive() {
+                        let missing = get_missing_variants(cx, arms, e);
+                        if !missing.is_empty() {
+                            let last = arms.last().unwrap().span;
+                            lint(cx, expr.span, last, missing)
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    fn check_pat(&mut self, cx: &LateContext<'tcx>, pat: &'tcx Pat<'_>) {
+        if let PatKind::Struct(_, fields @ [_, ..], true) = &pat.kind {
+            if let Some(ty::Adt(def, _)) = cx.maybe_typeck_results().map(|r| r.pat_ty(pat).kind()) {
+                if let Some(var) = def.variants.iter().next() {
+                    if var.is_field_list_non_exhaustive() {
+                        let missing = get_missing_fields(fields, &var.fields);
+                        if !missing.is_empty() {
+                            let last = fields.last().unwrap().span;
+                            lint(cx, pat.span, last, missing)
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+fn lint(cx: &LateContext<'_>, all: Span, _last: Span, missing: Vec<String>) {
+    cx.struct_span_lint(REACHABLE_PATTERNS, all, |lint: LintDiagnosticBuilder<'_>| {
+        let mut l = lint.build("missing patterns of non_exhaustive type");
+        l.help(&format!("add {} to match all reachable patterns", missing.join(",")));
+        l.emit();
+    });
+}
+
+fn get_missing_variants<'tcx>(
+    cx: &LateContext<'tcx>,
+    arms: &[Arm<'_>],
+    e: &'tcx Expr<'_>,
+) -> Vec<String> {
+    let ty = cx.typeck_results().expr_ty(e);
+    let mut missing_variants = vec![];
+    if let ty::Adt(def, _) = ty.kind() {
+        for variant in &def.variants {
+            missing_variants.push(variant);
+        }
+    }
+    let mut has_attr = false;
+    for arm in arms {
+        has_attr |= has_reachable_attr(cx, arm);
```

Comment:
It might be better to have the attribute on the match expression itself, it strikes me as unusual that the attribute is to be applied to *any* of the match arms to apply this lint.

Reply:
Ahh yeah, my logic is wrong. The idea, from the issue and discussions when `non_exhaustive` was impled, was that it only applies to a wildcard match `_`. I personally don't have much of a preference and could either move it to the expr/stmt position or fix the logic to reflect that it has to also be a wildcard either `_` for enums or `..` for structs this having an issue of it's own #81282.

Reply:
I think this will be decided either by the compiler team in the MCP, or the language team might be asked.